lib/iris/fileformats/grib/message.py

```diff
@@ -53,27 +62,28 @@ def messages_from_filename(filename):
             Name of the file to generate fields from.
 
         """
-        with open(filename, 'rb') as grib_fh:
-            while True:
-                offset = grib_fh.tell()
-                grib_id = gribapi.grib_new_from_file(grib_fh)
-                if grib_id is None:
-                    break
-                raw_message = _RawGribMessage(grib_id)
-                # ensure that all raw_message header Sections have
-                # their keys and values read from the file.
-                raw_message._cache_header()
-                recreate_raw = _MessageLocation(filename, offset)
-                yield GribMessage(raw_message, recreate_raw)
-
-    def __init__(self, raw_message, recreate_raw):
+        grib_fh = open(filename, 'rb')
+        file_ref = _OpenFileRef(grib_fh)
+
+        while True:
+            offset = grib_fh.tell()
+            grib_id = gribapi.grib_new_from_file(grib_fh)
+            if grib_id is None:
+                break
+            raw_message = _RawGribMessage(grib_id)
+            recreate_raw = _MessageLocation(filename, offset)
+            yield GribMessage(raw_message, recreate_raw)
```

Comment:
Important bit missing : should say "GribMessage(raw_message, recreate_raw**, file_ref=file_ref**)"